crux_time/src/lib.rs

```diff
@@ -44,33 +61,135 @@ where
             async move {
-                let response = context.request_from_shell(TimeRequest).await;
-                let response = response
-                    .0
-                    .parse::<DateTime<Utc>>()
-                    .expect("invalid time format");
-                context.update_app(callback(response));
+                context.update_app(callback(context.request_from_shell(TimeRequest::Now).await));
             }
         });
     }
 
-    /// Request current time, which will be returned as `chrono::DateTime<Utc>`.
+    /// Request current time, which will be passed to the app as a [`TimeResponse`] containing an [`Instant`]
     /// This is an async call to use with [`crux_core::compose::Compose`].
-    pub async fn now_async(&self) -> DateTime<Utc> {
+    pub async fn now_async(&self) -> TimeResponse {
+        self.context.request_from_shell(TimeRequest::Now).await
+    }
+
+    /// Subscribe to receive a notification when the specified [`Instant`] has arrived.
+    pub fn subscribe_instant<F>(&self, instant: Instant, callback: F)
+    where
+        F: Fn(TimeResponse) -> Ev + Send + Sync + 'static,
+    {
+        self.context.spawn({
+            let context = self.context.clone();
+            async move {
+                context.update_app(callback(
+                    context
+                        .request_from_shell(TimeRequest::SubscribeInstant(instant))
+                        .await,
+                ));
+            }
+        });
+    }
+
+    /// Subscribe to receive a notification when the specified [`Instant`] has arrived.
+    /// This is an async call to use with [`crux_core::compose::Compose`].
+    pub async fn subscribe_instant_async(&self, instant: Instant) -> TimeResponse {
         self.context
-            .request_from_shell(TimeRequest)
+            .request_from_shell(TimeRequest::SubscribeInstant(instant))
             .await
-            .0
-            .parse::<DateTime<Utc>>()
-            .expect("invalid time format")
+    }
+
+    /// Subscribe to receive a notification when the specified duration has elapsed.
+    pub fn subscribe_duration<F>(&self, duration: Duration, callback: F)
+    where
+        F: Fn(TimeResponse) -> Ev + Send + Sync + 'static,
+    {
+        self.context.spawn({
+            let context = self.context.clone();
+            async move {
+                context.update_app(callback(
+                    context
+                        .request_from_shell(TimeRequest::SubscribeDuration(duration))
+                        .await,
+                ));
+            }
+        });
+    }
+
+    /// Subscribe to receive a notification when the specified duration has elapsed.
+    /// This is an async call to use with [`crux_core::compose::Compose`].
+    pub async fn subscribe_duration_async(&self, duration: Duration) -> TimeResponse {
+        self.context
+            .request_from_shell(TimeRequest::SubscribeDuration(duration))
+            .await
+    }
```

Comment:
This function is often called sleep elsewhere - I wonder if borrowing that name here would be a good idea?